bookkeeper-server/src/main/java/org/apache/bookkeeper/discover/ZKRegistrationClient.java

```diff
@@ -133,13 +122,15 @@ public void safeRun() {
         @Override
         public void accept(Versioned<Set<BookieSocketAddress>> bookieSet, Throwable throwable) {
             if (throwable != null) {
-                scheduleWatchTask(ZK_CONNECT_BACKOFF_MS);
-                firstRunFuture.completeExceptionally(throwable);
+                if (firstRunFuture.isDone()) {
```

Comment:
```java
if (!firstRunFuture.completeExceptionally(throwable)) {
    scheduleWatchTask(ZK_CONNECT_BACKOFF_MS); 
}
```

But in any case, I don't get what you are doing with this code. Why not reschedule if firstRunFuture is complete?

Reply:
the register method return a completefuture. so I am aligning the behavior with the method signature. because if register failure, the task was removed, we don't need to retry this.

Reply:
ok, so the logic is basically, if the failure has been notified to the client, give up, otherwise retry.

Really we should be using something like observables here, but that's not for this change. In any case, make the check and completion atomic as above, since we're dealing with multiple threads here.

Reply:
yes that is aligned to the method signature, I don't change the original design here.